File: solution/backend/backend_springboot/src/main/java/it/unito/iumtweb/springboot/movies/Movies.java
package it.unito.iumtweb.springboot.movies;

import jakarta.persistence.*;
import java.io.Serializable;

/**
 * JPA Entity representing the 'movies' table.
 * <p>
 * This is the central entity of the system. It contains the main metadata for films.
 * It uses a standard numerical Primary Key (id) which is referenced by foreign keys in Actors, Crew, etc.
 * </p>
 */
@Entity
@Table(name = "movies")
public class Movies implements Serializable {

    // 1. Definiamo un generatore di sequenza con il nome standard di Postgres
    @SequenceGenerator(name = "movies_seq", sequenceName = "movies_id_seq", allocationSize = 1)

    // 2. Usiamo la strategia SEQUENCE.
    // IMPORTANTE: Usiamo Long invece di Integer per compatibilità con JpaRepository<Movies, Long>
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "movies_seq")
    private Long id;

    /** The title of the movie. */
    private String name;

    /**
     * The release year.
     * Mapped as Integer because the CSV/DB contains only the year (e.g., 2024).
     */
    @Column(name = "date")
    private Integer date;

    /** The promotional slogan of the movie. */
    private String tagline;

    /**
     * A short story abstract or plot summary.
     * Stored as TEXT to allow long descriptions.
     */
    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    /** The duration of the movie in minutes. */
    private Integer minute;

    /** The average rating score. */
    private Float rating;

    /**
     * Default constructor.
     */
    public Movies() {}

    /**
     * Full constructor.
     */
    public Movies(Long id, String name, Integer date, String tagline, String description, Integer minute, Float rating) {
        this.id = id;
        this.name = name;
        this.date = date;
        this.tagline = tagline;
        this.description = description;
        this.minute = minute;
        this.rating = rating;
    }

    // --- Getters and Setters ---

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public Integer getDate() { return date; }
    public void setDate(Integer date) { this.date = date; }

    public String getTagline() { return tagline; }
    public void setTagline(String tagline) { this.tagline = tagline; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public Integer getMinute() { return minute; }
    public void setMinute(Integer minute) { this.minute = minute; }

    public Float getRating() { return rating; }
    public void setRating(Float rating) { this.rating = rating; }
}

File: solution/backend/express_server/models/movies.js
const mongoose = require('mongoose');

/**
 * @swagger
 * components:
 *   schemas:
 *     Movie:
 *       type: object
 *       required:
 *         - name
 *       properties:
 *         name:
 *           type: string
 *           maxLength: 50
 *           description: Title of the movie
 *         year:
 *           type: integer
 *           description: Release year of the movie
 *         tagline:
 *           type: string
 *           maxLength: 100
 *           description: Promotional tagline
 *         description:
 *           type: string
 *           maxLength: 300
 *           description: Description of the movie
 *         minute:
 *           type: integer
 *           description: Duration of the movie in minutes
 *         rating:
 *           type: number
 *           description: Rating score of the movie
 */

/**
 * Mongoose schema for Movie documents.
 * Defines the structure of movie documents stored in MongoDB.
 */
const Movie = new mongoose.Schema({

    /** Unique identifier matching the CSV and Spring Boot DB (Long/Integer in SQL). */
    id: { type: Number, required: true, unique: true },

    /** Title of the movie (required, max length 50). */
    name: { type: String, required: true},

    /** Release year of the movie. */
    year: { type: Number },

    /** Promotional tagline (max length 100). */
    tagline: { type: String, max: 100 },

    /** Description of the movie (max length 300). */
    description: { type: String },

    /** Duration of the movie in minutes. */
    minute: { type: Number },

    /** Rating score of the movie. */
    rating: { type: Number }
});

/** Exports the Movie model based on the defined schema. */
module.exports = mongoose.model('movie', Movie);
